### PR TITLE
[ntuple] check for valid NTuple top-level field names (ROOT-10856)

### PR DESCRIPTION
**Pull request by mxxo**
Fixes ROOT-10856 by keeping track of field names and throwing an exception if there's a name conflict. 

Edit: expanded PR scope to ensure field names are valid as well as unique (e.g. no empty string).  

~~**todo**~~ **done:** add unit test per method to ensure technique works:
- [x] `std::shared_ptr<T> MakeField(std::string_view fieldName, ArgsT&&... args)`
- [x] `void AddField(std::string_view fieldName, T* fromWhere)`
- [x] `void AddField(std::unique_ptr<Detail::RFieldBase> field)`
- [x] `std::shared_ptr<RCollectionNTuple> MakeCollection(std::string_view fieldName, std::unique_ptr<RNTupleModel> collectionModel)`

### CODE REVIEW

File: tree/ntuple/v7/src/RNTupleModel.cxx [hunk 1]
```diff
@@ -20,10 +20,18 @@
...
 #include <cstdlib>
+#include <exception>
 #include <memory>
 #include <utility>
 
 
+void ROOT::Experimental::RNTupleModel::EnsureUniqueFieldName(const std::string& fieldName)
+{
+   if (fFieldNames.insert(fieldName).second == false) {
+      throw std::runtime_error("RNTupleModel::MakeField: field name '" + fieldName + "' already exists");
```
Comment (jblomer) [inline tree/ntuple/v7/src/RNTupleModel.cxx:31]: Let's throw a ROOT specific `RException`

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx [hunk 2]
```diff
@@ -49,6 +50,10 @@ class RNTupleModel {
    std::unique_ptr<RFieldRoot> fRootField;
    /// Contains field values corresponding to the created top-level fields
    std::unique_ptr<REntry> fDefaultEntry;
+   /// Keeps track of which field names are taken.
+   std::unordered_set<std::string> fFieldNames;
+
+   void EnsureUniqueFieldName(const std::string& fieldName);
```
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx:56]: use `std::string_view`

File: tree/ntuple/v7/test/ntuple_basics.cxx [hunk 3]
```diff
@@ -211,3 +211,19 @@ TEST(RNTuple, Clusters)
...
+
+TEST(RNTupleModel, EnforceUniqueFieldNames)
+{
+   auto model = RNTupleModel::Create();
+   auto field = model->MakeField<float>("pt", 42.0);
+   try {
+      auto field2 = model->MakeField<float>("pt", 42.0);
+      FAIL() << "repeated field names should throw";
+   } catch (const std::runtime_error& err) {
+     EXPECT_STREQ(err.what(), "RNTupleModel::MakeField: field name 'pt' already exists");
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_basics.cxx:223]: The `RException` might surround the original error message with more text

File: tree/ntuple/v7/test/ntuple_basics.cxx [hunk 4]
```diff
@@ -211,3 +211,19 @@ TEST(RNTuple, Clusters)
...
+   }
+   auto field2 = model->MakeField<float>("pt2", 42.0);
+
+   // corner case -- doesn't apply to NTuple's FieldZero
+   auto field3 = model->MakeField<float>("", 42.0);
```
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_basics.cxx:228]: That should not be possible, either, I think.  Field names must not be empty.

File: tree/ntuple/v7/src/RNTupleModel.cxx (hunk 1 above)
Comment (mxxo) [reply tree/ntuple/v7/src/RNTupleModel.cxx]: Thanks, done in eea9c7c74fdeb2119d9b2949cd5b9fe6e6d13edc.

File: tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx (hunk 2 above)
Comment (mxxo) [reply tree/ntuple/v7/inc/ROOT/RNTupleModel.hxx]: Done in 49624995939ab0ee78c73384c5d65e83b343ce49.

File: tree/ntuple/v7/test/ntuple_basics.cxx (hunk 4 above)
Comment (mxxo) [reply tree/ntuple/v7/test/ntuple_basics.cxx]: I added a check in  03ddbb3384e4378358304a4bcf121866

File: tree/ntuple/v7/src/RNTupleModel.cxx [hunk 5]
```diff
@@ -13,17 +13,28 @@
...
 
+void ROOT::Experimental::RNTupleModel::EnsureValidFieldName(std::string_view fieldName)
+{
+   auto fieldNameStr = std::string(fieldName);
+   if (fFieldNames.insert(fieldNameStr).second == false) {
+      throw RException(R__FAIL("field name '" + fieldNameStr + "' already exists"));
+   } else if (fieldNameStr == "") {
+      throw RException(R__FAIL("field name cannot be empty string \"\""));
+   } else if (fieldNameStr.find(".") != std::string::npos) {
+      throw RException(R__FAIL("field name '" + fieldNameStr + "' cannot contain periods '.'"));
```
Comment (mxxo) [inline tree/ntuple/v7/src/RNTupleModel.cxx:34]: Is period OK here? Maybe dot instead?

File: tree/ntuple/v7/test/ntuple_test.hxx [hunk 6]
```diff
@@ -26,6 +27,7 @@
 #include <TFile.h>
 #include <TRandom3.h>
 
+#include "gmock/gmock.h"
```
Comment (mxxo) [inline tree/ntuple/v7/test/ntuple_test.hxx:30]: I brought in gmock (part of googletest) to help with substring checking.

File: tree/ntuple/v7/test/ntuple_basics.cxx [hunk 7]
```diff
@@ -211,3 +211,60 @@ TEST(RNTuple, Clusters)
...
+   // MakeCollection
+   try {
+      auto otherModel = RNTupleModel::Create();
+      auto collection = model->MakeCollection("pt", std::move(otherModel));
+      FAIL() << "repeated field names should throw";
+   } catch (const RException& err) {
+      EXPECT_THAT(err.what(), testing::HasSubstr("field name 'pt' already exists"));
+   }
+
+   // subfield names don't throw because full name differs (otherModel.pt)
```
Comment (mxxo) [inline tree/ntuple/v7/test/ntuple_basics.cxx:266]: On my first attempt, I checked for duplicate sub-field names but this failed when repeated collection fields were added: 
```shell
C++ exception with description "field name 'std::vector<float>' already exists
```
I now think this the right approach to only check the collection field name and not the subfield names because their full names are different anyway: `otherModel.pt` when `otherModel` is added to the NTuple.

File: tree/ntuple/v7/src/RNTupleModel.cxx (hunk 5 above)
Comment (mxxo) [reply tree/ntuple/v7/src/RNTupleModel.cxx]: Done in 6372ea9
